### PR TITLE
Forward errors through /complete-auth endpoint – WEB-50

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Description
This PR improves reporting of errors which might happen during the OAuth/OIDC flow on backend.

Issues revealed:
* errors are reported twice, in some cases
* toast notification expires after a couple of seconds, user might miss them

Screenshot 2023-05-26 at 11 01 31

## Related Issue(s)

Fixes WEB-50

## How to test
* create any SSO config, e.g. using `accounts.google.com`, set an invalid secret and try to verify

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - at-forward-errors</li>
	<li><b>🔗 URL</b> - <a href="https://at-forward-errors.preview.gitpod-dev.com/workspaces" target="_blank">at-forward-errors.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [x] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>
 
/hold

### CODE REVIEW

File: components/public-api-server/pkg/oidc/oauth2.go [hunk 1]
```diff
@@ -73,17 +74,21 @@ func (s *Service) OAuth2Middleware(next http.Handler) http.Handler {
 			http.Error(rw, "bad state param", http.StatusBadRequest)
 			return
 		}
+		useHttpErrors := state.UseHttpErrors
```
Comment (easyCZ) [inline components/public-api-server/pkg/oidc/oauth2.go:77]: Can we not do this based on the ContentType which is being requested on the `/callback` endpoint?

Comment (easyCZ) [reply components/public-api-server/pkg/oidc/oauth2.go:77]: An in general, what prevents us from always responding with the appropriate HTTP status code, but use the ContentType to determine whether to respond with JSON, or text?

Comment (AlexTugarev) [reply components/public-api-server/pkg/oidc/oauth2.go:77]: Simple answer: the callback route isn't called by us, thus we cannot rely on HTTP headers to decide on that. 

Other way around, the fact that we're not interested in HTTP status code makes me think if we could drop that here.

File: components/public-api-server/pkg/oidc/router.go [hunk 2]
```diff
@@ -161,3 +171,23 @@ func (s *Service) getCallbackHandler() http.HandlerFunc {
 		http.Redirect(rw, r, oauth2Result.ReturnToURL, http.StatusTemporaryRedirect)
 	}
 }
+
+func respondeWithError(w http.ResponseWriter, r *http.Request, error string, code int, useHttpErrors bool) {
+	if useHttpErrors {
+		http.Error(w, error, code)
+	}
+	if !useHttpErrors {
```
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply components/public-api-server/pkg/oidc/router.go]: Thanks!

**Review by geropl — APPROVED** [review]
Code LGTM, tested and works. :heavy_check_mark: 

Ack'd [this discussion](https://github.com/gitpod-io/gitpod/pull/17753/files#r1206572413), but I'm lagging context, and think we should unblock Dedicated first. :lotus_position:

**Pull request merged**